File: src/otdev/tools/_arch/frontend/src/solution/projection.ts
import type {
  ArchitectureLevel,
  AbsentSelectedSystem,
  BoundaryInterface,
  CollapsedInterface,
  PreparedSolutionSnapshots,
  ProjectionDiagnostic,
  SystemSetSelector,
  ViewGraph,
  ViewGraphEdge,
  ViewGraphNode,
} from '../data/types'
import {
  BoundedCache,
  LAYOUT_SCHEMA_VERSION,
  PROJECTION_SCHEMA_VERSION,
  RENDERER_ADAPTER_VERSION,
} from './runtime'

export interface LocalSolutionProjection {
  cacheKey: string
  graph: ViewGraph
  selectedSystems: string[]
  includedSystems: string[]
  systemDistances: Record<string, number>
  absentSystems: AbsentSelectedSystem[]
  internalInterfaces: ViewGraphEdge[]
  boundaryInterfaces: BoundaryInterface[]
  collapsedInterfaces: CollapsedInterface[]
  diagnostics: ProjectionDiagnostic[]
}

export function normalizedSelector(selector: SystemSetSelector): SystemSetSelector {
  return {
    systems: [...new Set(selector.systems)].sort(),
    system_groups: [...new Set(selector.system_groups)].sort(),
    changes: [...new Set(selector.changes)].sort(),
    change_groups: [...new Set(selector.change_groups)].sort(),
    tags: [...new Set(selector.tags)].sort(),
  }
}

export function stableSelector(selector: SystemSetSelector): string {
  return JSON.stringify(normalizedSelector(selector))
}

const projectionCaches = new WeakMap<
  PreparedSolutionSnapshots,
  BoundedCache<string, LocalSolutionProjection>
>()

function projectionCache(prepared: PreparedSolutionSnapshots) {
  let cache = projectionCaches.get(prepared)
  if (!cache) {
    cache = new BoundedCache<string, LocalSolutionProjection>(32)
    projectionCaches.set(prepared, cache)
  }
  return cache
}

export function projectionCacheStats(prepared: PreparedSolutionSnapshots) {
  const cache = projectionCache(prepared)
  return { size: cache.size, hits: cache.hits, misses: cache.misses }
}

export function solutionProjectionCacheKey({
  prepared,
  snapshot,
  order,
  selector,
  interfaceDepth,
  level,
  theme,
}: {
  prepared: PreparedSolutionSnapshots
  snapshot: ViewGraph
  order: number
  selector: SystemSetSelector
  interfaceDepth: number
  level: ArchitectureLevel
  theme: string
}): string {
  return [
    PROJECTION_SCHEMA_VERSION,
    RENDERER_ADAPTER_VERSION,
    LAYOUT_SCHEMA_VERSION,
    prepared.roadmap_id,
    snapshot.id,
    order,
    stableSelector(selector),
    interfaceDepth,
    level,
    theme,
  ].join('|')
}

function stableHash(value: string): string {
  let hash = 2166136261
  for (let index = 0; index < value.length; index += 1) {
    hash ^= value.charCodeAt(index)
    hash = Math.imul(hash, 16777619)
  }
  return (hash >>> 0).toString(16).padStart(8, '0')
}

function resolveSystems(
  prepared: PreparedSolutionSnapshots,
  order: number,
  selector: SystemSetSelector,
): Set<string> {
  const indexes = prepared.indexes[String(order)]
  if (!indexes) return new Set()
  const lookups: [string, string[], Set<string>][] = [
    ['systems', selector.systems, new Set(indexes.systems)],
    ['system groups', selector.system_groups, new Set(Object.keys(indexes.system_groups))],
    ['changes', selector.changes, new Set(Object.keys(indexes.changes))],
    ['change groups', selector.change_groups, new Set(Object.keys(indexes.change_groups))],
    ['tags', selector.tags, new Set(Object.keys(indexes.tags))],
  ]
  for (const [label, values, known] of lookups) {
    const unknown = values.filter((value) => !known.has(value)).sort()
    if (unknown.length > 0) throw new Error(`Unknown ${label}: ${unknown.join(', ')}`)
  }
  const requested = Object.values(selector).some((values) => values.length > 0)
  const selected = new Set(selector.systems)
  for (const group of selector.system_groups)
    indexes.system_groups[group]?.forEach((id) => selected.add(id))
  for (const change of selector.changes)
    indexes.changes[change]?.forEach((id) => selected.add(id))
  for (const group of selector.change_groups)
    indexes.change_groups[group]?.forEach((id) => selected.add(id))
  for (const tag of selector.tags) indexes.tags[tag]?.forEach((id) => selected.add(id))
  return requested ? selected : new Set(indexes.systems)
}

function systemMap(nodes: ViewGraphNode[]): Map<string, string | undefined> {
  const byId = new Map(nodes.map((node) => [node.id, node]))
  const result = new Map<string, string | undefined>()
  for (const node of nodes) {
    let current: ViewGraphNode | undefined = node
    const seen = new Set<string>()
    while (current && current.entity_kind !== 'system') {
      if (!current.parent || seen.has(current.id)) {
        current = undefined
        break
      }
      seen.add(current.id)
      current = byId.get(current.parent)
    }
    result.set(node.id, current?.id)
  }
  return result
}

function levelEndpoint(
  endpoint: string,
  level: ArchitectureLevel,
  nodes: Map<string, ViewGraphNode>,
): string | undefined {
  let current = nodes.get(endpoint)
  while (current) {
    if (level === 'component' || current.entity_kind === 'system') return current.id
    if (level === 'application' && current.entity_kind === 'application') return current.id
    current = current.parent ? nodes.get(current.parent) : undefined
  }
  return undefined
}

export function projectSolution(
  prepared: PreparedSolutionSnapshots,
  order: number,
  selector: SystemSetSelector,
  interfaceDepth: number,
  level: ArchitectureLevel,
  theme = 'clean',
): LocalSolutionProjection | undefined {
  if (!Number.isInteger(interfaceDepth) || interfaceDepth < 0)
    throw new Error('interfaceDepth must be a non-negative integer')
  const snapshot = prepared.snapshots[String(order)]
  if (!snapshot) return undefined
  const resolvedSelector = normalizedSelector(selector)
  const cacheKey = solutionProjectionCacheKey({
    prepared,
    snapshot,
    order,
    selector: resolvedSelector,
    interfaceDepth,
    level,
    theme,
  })
  const cached = projectionCache(prepared).get(cacheKey)
  if (cached) return cached
  const selected = resolveSystems(prepared, order, resolvedSelector)
  const systems = systemMap(snapshot.nodes)
  const presentSystems = new Set(
    snapshot.nodes
      .filter(
        (node) =>
          node.entity_kind === 'system' && !node.tombstone && !node.future,
      )
      .map((node) => node.id),
  )
  const absentSystems = [...selected]
    .filter((systemId) => !presentSystems.has(systemId))
    .sort()
    .map((systemId): AbsentSelectedSystem => {
      const presence = prepared.system_presence[systemId] ?? []
      const state =
        presence.length > 0 && order < Math.min(...presence)
          ? 'not_yet_present'
          : presence.length > 0 && order > Math.max(...presence)
            ? 'no_longer_present'
            : 'not_present'
      return { system_id: systemId, state, message: 'not present at this snapshot' }
    })
  const adjacency = new Map<string, Set<string>>()
  for (const edge of snapshot.edges) {
    if (edge.entity_kind !== 'interface') continue
    const source = systems.get(edge.source_id)
    const target = systems.get(edge.target_id)
    if (!source || !target || source === target) continue
    if (!adjacency.has(source)) adjacency.set(source, new Set())
    if (!adjacency.has(target)) adjacency.set(target, new Set())
    adjacency.get(source)!.add(target)
    adjacency.get(target)!.add(source)
  }
  const distances = new Map([...selected].map((id): [string, number] => [id, 0]))
  const included = new Set(distances.keys())
  let frontier = new Set(selected)
  for (let hop = 0; hop < interfaceDepth; hop += 1) {
    const next = new Set<string>()
    for (const system of frontier) {
      for (const neighbor of adjacency.get(system) ?? []) {
        if (!included.has(neighbor)) next.add(neighbor)
      }
    }
    next.forEach((id) => included.add(id))
    next.forEach((id) => distances.set(id, hop + 1))
    frontier = next
  }

  const internal: ViewGraphEdge[] = []
  const boundaryInterfaces: BoundaryInterface[] = []
  const diagnostics: ProjectionDiagnostic[] = []
  const snapshotNodeById = new Map(snapshot.nodes.map((node) => [node.id, node]))
  for (const edge of snapshot.edges) {
    const sourceSystem = systems.get(edge.source_id)
    const targetSystem = systems.get(edge.target_id)
    for (const [endpointId, endpointSystem] of [
      [edge.source_id, sourceSystem],
      [edge.target_id, targetSystem],
    ] as const) {
      const endpoint = snapshotNodeById.get(endpointId)
      if (!endpointSystem && endpoint?.entity_kind !== 'user') {
        diagnostics.push({
          code:
            edge.entity_kind === 'interface'
              ? 'unresolved_interface_endpoint'
              : 'unresolved_relationship_endpoint',
          message: `${edge.entity_kind} '${edge.id}' endpoint '${endpointId}' has no owning system`,
          entity_id: edge.id,
          endpoint_id: endpointId,
        })
      }
    }
    const sourceInside = included.has(sourceSystem ?? '')
    const targetInside = included.has(targetSystem ?? '')
    if (sourceInside && targetInside) internal.push(edge)
    else if (edge.entity_kind === 'interface' && sourceInside !== targetInside)
      boundaryInterfaces.push({
        interface: edge,
        inside_system: (sourceInside ? sourceSystem : targetSystem)!,
        inside_endpoint: sourceInside ? edge.source_id : edge.target_id,
        outside_system: sourceInside ? targetSystem : sourceSystem,
        outside_endpoint: sourceInside ? edge.target_id : edge.source_id,
      })
  }

  const allowed = {
    system: new Set(['system']),
    application: new Set(['system', 'application']),
    component: new Set(['system', 'application', 'component']),
  }[level]
  const nodes = snapshot.nodes.filter(
    (node) => allowed.has(node.entity_kind) && included.has(systems.get(node.id) ?? ''),
  )
  const nodeById = new Map(nodes.map((node) => [node.id, node]))
  const allNodes = new Map(snapshot.nodes.map((node) => [node.id, node]))
  const children = new Map<string, string[]>()
  for (const node of nodes) {
    if (!node.parent || !nodeById.has(node.parent)) continue
    children.set(node.parent, [...(children.get(node.parent) ?? []), node.id])
  }
  const projectedNodes = nodes.map((node) => ({
    ...node,
    children: [...(children.get(node.id) ?? [])].sort(),
  }))
  const collapsedInterfaces: CollapsedInterface[] = []
  const relationships: ViewGraphEdge[] = []
  const interfaceGroups = new Map<string, ViewGraphEdge[]>()
  for (const edge of internal) {
    const source = levelEndpoint(edge.source_id, level, allNodes)
    const target = levelEndpoint(edge.target_id, level, allNodes)
    if (!source || !target || !nodeById.has(source) || !nodeById.has(target)) continue
    const projected = { ...edge, source_id: source, target_id: target }
    if (edge.entity_kind === 'interface' && source === target) {
      collapsedInterfaces.push({
        interface: edge,
        visible_node: source,
        reason: 'collapsed_within_visible_node',
      })
      continue
    }
    if (edge.entity_kind === 'relationship') {
      relationships.push(projected)
      continue
    }
    const key = [
      source,
      target,
      edge.direction,
      edge.integration_type ?? '',
      edge.status,
      edge.context_status,
    ].join('|')
    interfaceGroups.set(key, [...(interfaceGroups.get(key) ?? []), projected])
  }
  const edges = [...relationships]
  for (const [key, values] of [...interfaceGroups].sort(([left], [right]) => left.localeCompare(right))) {
    const members = [...values].sort((left, right) => left.id.localeCompare(right.id))
    if (members.length === 1) {
      edges.push(members[0]!)
      continue
    }
    const first = members[0]!
    const memberIds = members.map((item) => item.id)
    edges.push({
      ...first,
      id: `aggregate-${stableHash([key, ...memberIds].join('|'))}`,
      name: `${members.length} interfaces`,
      description: undefined,
      interface_ids: memberIds,
      tags: [...new Set(members.flatMap((item) => item.tags))].sort(),
      related_changes: [...new Set(members.flatMap((item) => item.related_changes))].sort(),
      properties: { aggregate_members: memberIds },
    })
  }
  edges.sort((left, right) => left.id.localeCompare(right.id))
  const result: LocalSolutionProjection = {
    cacheKey,
    selectedSystems: [...selected].sort(),
    includedSystems: [...included].sort(),
    systemDistances: Object.fromEntries([...distances].sort(([left], [right]) => left.localeCompare(right))),
    absentSystems,
    internalInterfaces: internal
      .filter((edge) => edge.entity_kind === 'interface')
      .sort((left, right) => left.id.localeCompare(right.id)),
    boundaryInterfaces,
    collapsedInterfaces: collapsedInterfaces.sort((left, right) => left.interface.id.localeCompare(right.interface.id)),
    diagnostics: diagnostics.sort((left, right) =>
      `${left.entity_id}|${left.endpoint_id}|${left.code}`.localeCompare(
        `${right.entity_id}|${right.endpoint_id}|${right.code}`,
      ),
    ),
    graph: {
      ...snapshot,
      id: `solution-${stableHash(cacheKey)}`,
      selection: {
        ...snapshot.selection,
        selection: {
          ...snapshot.selection.selection,
          system_set: resolvedSelector,
          interface_depth: interfaceDepth,
          level,
          theme,
        },
      },
      nodes: projectedNodes,
      containers: [...children.keys()].sort(),
      edges,
    },
  }
  projectionCache(prepared).set(cacheKey, result)
  return result
}

File: src/otdev/tools/_arch/v2/projection.py
"""Snapshot preparation and deterministic local solution projections."""

from __future__ import annotations

import hashlib
import json
from collections import defaultdict
from typing import Literal, cast

from .models import (
    AbsentSelectedSystem,
    ArchitectureLevel,
    ArchitectureWorkspace,
    BoundaryInterface,
    CollapsedInterface,
    PreparedSolutionSnapshots,
    ProjectionDiagnostic,
    SolutionProjection,
    SystemSetSelector,
    ViewGraph,
    ViewGraphEdge,
    ViewGraphNode,
    ViewSelection,
)
from .replay import RoadmapReplayTimeline, prepare_roadmap_timeline
from .selection import selection_identity
from .selectors import SelectorError, roadmap_selection_indexes, systems_for_selector
from .viewgraph import resolve_view_graph

_PROJECTION_SCHEMA_VERSION = "solution-projection-2"
_RENDERER_ADAPTER_VERSION = "likec4-adapter-1"
_LAYOUT_SCHEMA_VERSION = "solution-layout-1"


class SolutionProjectionError(ValueError):
    """Raised when a snapshot or selector cannot be resolved."""


def _stable_hash(value: str) -> str:
    result = 2_166_136_261
    for character in value:
        result ^= ord(character)
        result = (result * 16_777_619) & 0xFFFFFFFF
    return f"{result:08x}"


def prepare_solution_snapshots(
    *,
    workspace: ArchitectureWorkspace,
    roadmap_id: str,
    replay_timeline: RoadmapReplayTimeline | None = None,
) -> PreparedSolutionSnapshots:
    """Precompute validated full graphs and indexes, but no projection layouts."""
    roadmap = next((item for item in workspace.roadmaps if item.id == roadmap_id), None)
    if roadmap is None:
        raise SolutionProjectionError(f"Unknown roadmap '{roadmap_id}'")
    orders = [
        0,
        *(item.order for item in sorted(roadmap.items, key=lambda item: item.order)),
    ]
    snapshots: dict[str, ViewGraph] = {}
    unavailable: list[int] = []
    timeline = replay_timeline or prepare_roadmap_timeline(
        workspace=workspace,
        roadmap_id=roadmap_id,
    )
    for order in orders:
        selection: dict[str, object] = {"roadmap": roadmap_id, "order": order}
        if order > 0:
            selection["compare_from"] = order - 1
        resolved = resolve_view_graph(
            workspace=workspace,
            value=selection,
            replay_timeline=timeline,
        )
        if resolved.graph is None or resolved.issues.errors:
            unavailable.append(order)
            continue
        key = str(order)
        snapshots[key] = resolved.graph
    roadmap_indexes = roadmap_selection_indexes(
        workspace=workspace,
        roadmap_id=roadmap_id,
        timeline=timeline,
    )
    presence: dict[str, list[int]] = defaultdict(list)
    for key, graph in snapshots.items():
        for node in graph.nodes:
            if node.entity_kind == "system" and not node.tombstone and not node.future:
                presence[node.id].append(int(key))
    return PreparedSolutionSnapshots(
        roadmap_id=roadmap_id,
        snapshots=snapshots,
        indexes={key: roadmap_indexes.model_copy(deep=True) for key in snapshots},
        system_presence={key: sorted(value) for key, value in sorted(presence.items())},
        unavailable_orders=unavailable,
    )


def _normalized_selector(selector: SystemSetSelector) -> SystemSetSelector:
    return selector.model_copy(
        update={
            key: sorted(set(values))
            for key, values in selector.model_dump(mode="python").items()
        }
    )


def _system_map(nodes: list[ViewGraphNode]) -> dict[str, str | None]:
    by_id = {node.id: node for node in nodes}
    result: dict[str, str | None] = {}
    for node in nodes:
        current: ViewGraphNode | None = node
        seen: set[str] = set()
        while current is not None and current.entity_kind != "system":
            if current.id in seen or current.parent is None:
                current = None
                break
            seen.add(current.id)
            current = by_id.get(current.parent)
        result[node.id] = current.id if current is not None else None
    return result


def _system_distances(
    *, graph: ViewGraph, selected: set[str], depth: int, systems: dict[str, str | None]
) -> dict[str, int]:
    adjacency: dict[str, set[str]] = defaultdict(set)
    for edge in graph.edges:
        if edge.entity_kind != "interface":
            continue
        source = systems.get(edge.source_id)
        target = systems.get(edge.target_id)
        if source is None or target is None or source == target:
            continue
        adjacency[source].add(target)
        adjacency[target].add(source)
    distances = dict.fromkeys(selected, 0)
    included = set(distances)
    frontier = set(selected)
    for hop in range(1, depth + 1):
        frontier = {
            neighbor for item in frontier for neighbor in adjacency[item]
        } - included
        included.update(frontier)
        distances.update(dict.fromkeys(frontier, hop))
    return distances


def _level_endpoint(
    *, endpoint: str, level: ArchitectureLevel, nodes: dict[str, ViewGraphNode]
) -> str | None:
    current = nodes.get(endpoint)
    while current is not None:
        if level == "component" or current.entity_kind == "system":
            return current.id
        if level == "application" and current.entity_kind == "application":
            return current.id
        current = nodes.get(current.parent) if current.parent else None
    return None


def solution_projection_cache_key(
    *,
    snapshot_id: str,
    model_id: str,
    selector: SystemSetSelector,
    depth: int,
    level: ArchitectureLevel,
    theme: str,
) -> str:
    """Return the stable cache identity; coloring is deliberately excluded."""
    payload = {
        "snapshot": snapshot_id,
        "model": model_id,
        "selector": {
            key: sorted(set(values))
            for key, values in selector.model_dump(mode="json").items()
        },
        "depth": depth,
        "level": level,
        "theme": theme,
        "versions": {
            "projection": _PROJECTION_SCHEMA_VERSION,
            "renderer": _RENDERER_ADAPTER_VERSION,
            "layout": _LAYOUT_SCHEMA_VERSION,
        },
    }
    encoded = json.dumps(payload, separators=(",", ":"), sort_keys=True).encode()
    return hashlib.sha256(encoded).hexdigest()


def project_solution(
    *,
    prepared: PreparedSolutionSnapshots,
    order: int,
    selector: SystemSetSelector | dict[str, object] | None = None,
    interface_depth: int = 0,
    level: ArchitectureLevel = "system",
    selection: ViewSelection | None = None,
) -> SolutionProjection:
    """Derive one projection from a prepared snapshot without resolving the roadmap again."""
    key = str(order)
    graph = prepared.snapshots.get(key)
    indexes = prepared.indexes.get(key)
    if graph is None or indexes is None:
        raise SolutionProjectionError(
            f"Roadmap '{prepared.roadmap_id}' order {order} is unavailable"
        )
    if interface_depth < 0:
        raise SolutionProjectionError("interface_depth must be at least zero")
    resolved_selector = _normalized_selector(
        selector
        if isinstance(selector, SystemSetSelector)
        else SystemSetSelector.model_validate(selector or {})
    )
    try:
        selected = systems_for_selector(selector=resolved_selector, indexes=indexes)
    except SelectorError as exc:
        raise SolutionProjectionError(str(exc)) from exc
    system_by_node = _system_map(graph.nodes)
    distances = _system_distances(
        graph=graph,
        selected=selected,
        depth=interface_depth,
        systems=system_by_node,
    )
    included = set(distances)
    present_systems = {
        node.id
        for node in graph.nodes
        if node.entity_kind == "system" and not node.tombstone and not node.future
    }
    absent_systems: list[AbsentSelectedSystem] = []
    for system_id in sorted(selected - present_systems):
        presence = prepared.system_presence.get(system_id, [])
        if presence and order < min(presence):
            state = "not_yet_present"
        elif presence and order > max(presence):
            state = "no_longer_present"
        else:
            state = "not_present"
        absent_systems.append(
            AbsentSelectedSystem(
                system_id=system_id,
                state=cast(
                    "Literal['not_yet_present', 'no_longer_present', 'not_present']",
                    state,
                ),
            )
        )

    internal_edges: list[ViewGraphEdge] = []
    boundary: list[BoundaryInterface] = []
    diagnostics: list[ProjectionDiagnostic] = []
    graph_nodes = {node.id: node for node in graph.nodes}
    for edge in graph.edges:
        source_system = system_by_node.get(edge.source_id)
        target_system = system_by_node.get(edge.target_id)
        for endpoint_id, endpoint_system in (
            (edge.source_id, source_system),
            (edge.target_id, target_system),
        ):
            endpoint = graph_nodes.get(endpoint_id)
            if endpoint_system is None and (
                endpoint is None or endpoint.entity_kind != "user"
            ):
                diagnostics.append(
                    ProjectionDiagnostic(
                        code=(
                            "unresolved_interface_endpoint"
                            if edge.entity_kind == "interface"
                            else "unresolved_relationship_endpoint"
                        ),
                        message=(
                            f"{edge.entity_kind.title()} '{edge.id}' endpoint "
                            f"'{endpoint_id}' has no owning system"
                        ),
                        entity_id=edge.id,
                        endpoint_id=endpoint_id,
                    )
                )
        source_inside = source_system in included
        target_inside = target_system in included
        if source_inside and target_inside:
            internal_edges.append(edge)
        elif edge.entity_kind == "interface" and source_inside != target_inside:
            boundary.append(
                BoundaryInterface(
                    interface=edge,
                    inside_system=cast(
                        "str", source_system if source_inside else target_system
                    ),
                    inside_endpoint=edge.source_id if source_inside else edge.target_id,
                    outside_system=target_system if source_inside else source_system,
                    outside_endpoint=edge.target_id
                    if source_inside
                    else edge.source_id,
                )
            )

    allowed_kinds = {
        "system": {"system"},
        "application": {"system", "application"},
        "component": {"system", "application", "component"},
    }[level]
    projected_nodes = [
        node
        for node in graph.nodes
        if node.entity_kind in allowed_kinds and system_by_node.get(node.id) in included
    ]
    projected_by_id = {node.id: node for node in projected_nodes}
    children: dict[str, list[str]] = defaultdict(list)
    for node in projected_nodes:
        if node.parent is not None and node.parent in projected_by_id:
            children[node.parent].append(node.id)
    projected_nodes = [
        node.model_copy(update={"children": sorted(children[node.id])})
        for node in projected_nodes
    ]
    original_nodes = {node.id: node for node in graph.nodes}
    projected_relationships: list[ViewGraphEdge] = []
    projected_interfaces: dict[
        tuple[str, str, str, str | None, str, str], list[ViewGraphEdge]
    ] = defaultdict(list)
    collapsed: list[CollapsedInterface] = []
    for edge in internal_edges:
        source = _level_endpoint(
            endpoint=edge.source_id, level=level, nodes=original_nodes
        )
        target = _level_endpoint(
            endpoint=edge.target_id, level=level, nodes=original_nodes
        )
        if (
            source is None
            or target is None
            or source not in projected_by_id
            or target not in projected_by_id
        ):
            continue
        projected = edge.model_copy(update={"source_id": source, "target_id": target})
        if edge.entity_kind == "interface" and source == target:
            collapsed.append(CollapsedInterface(interface=edge, visible_node=source))
            continue
        if edge.entity_kind == "relationship":
            projected_relationships.append(projected)
            continue
        projected_interfaces[
            (
                source,
                target,
                edge.direction,
                edge.integration_type,
                edge.status,
                edge.context_status,
            )
        ].append(projected)

    projected_edges: list[ViewGraphEdge] = list(projected_relationships)
    for aggregate_key, members in sorted(projected_interfaces.items()):
        ordered = sorted(members, key=lambda item: item.id)
        if len(ordered) == 1:
            projected_edges.append(ordered[0])
            continue
        member_ids = [item.id for item in ordered]
        aggregate_identity = "|".join(
            [
                aggregate_key[0],
                aggregate_key[1],
                aggregate_key[2],
                aggregate_key[3] or "",
                aggregate_key[4],
                aggregate_key[5],
                *member_ids,
            ]
        )
        first = ordered[0]
        projected_edges.append(
            first.model_copy(
                update={
                    "id": f"aggregate-{_stable_hash(aggregate_identity)}",
                    "name": f"{len(ordered)} interfaces",
                    "description": None,
                    "interface_ids": member_ids,
                    "tags": sorted({tag for item in ordered for tag in item.tags}),
                    "related_changes": sorted(
                        {
                            change_id
                            for item in ordered
                            for change_id in item.related_changes
                        }
                    ),
                    "properties": {"aggregate_members": member_ids},
                }
            )
        )
    projected_edges.sort(key=lambda item: item.id)

    snapshot_id = f"{prepared.roadmap_id}@{order}"
    requested_selection = selection or graph.selection.selection
    projected_selection = requested_selection.model_copy(
        update={
            "system_set": resolved_selector,
            "interface_depth": interface_depth,
            "level": level,
        }
    )
    projected_selection_id = selection_identity(projected_selection)
    projected_graph = graph.model_copy(
        update={
            "id": f"solution-{projected_selection_id.removeprefix('selection-')}",
            "selection": graph.selection.model_copy(
                update={
                    "id": projected_selection_id,
                    "selection": projected_selection,
                }
            ),
            "nodes": projected_nodes,
            "containers": sorted(children),
            "edges": projected_edges,
        }
    )
    return SolutionProjection(
        cache_key=solution_projection_cache_key(
            snapshot_id=snapshot_id,
            model_id=graph.id,
            selector=resolved_selector,
            depth=interface_depth,
            level=level,
            theme=projected_selection.theme or "clean",
        ),
        snapshot_id=snapshot_id,
        selector=resolved_selector,
        selected_systems=sorted(selected),
        included_systems=sorted(included),
        system_distances={key: distances[key] for key in sorted(distances)},
        absent_systems=absent_systems,
        interface_depth=interface_depth,
        level=level,
        graph=projected_graph,
        internal_interfaces=[
            edge for edge in internal_edges if edge.entity_kind == "interface"
        ],
        boundary_interfaces=boundary,
        collapsed_interfaces=sorted(collapsed, key=lambda item: item.interface.id),
        diagnostics=sorted(
            diagnostics,
            key=lambda item: (item.entity_id, item.endpoint_id, item.code),
        ),
    )
